stash_engine/spec/db/stash_engine/identifier_spec.rb

```diff
@@ -469,6 +469,197 @@ module StashEngine
 
     end
 
+    describe '#calculated_pub_state' do
+
+      before(:each) do
+        # a way to neuter all the callback activity
+        allow_any_instance_of(CurationActivity).to receive(:update_solr).and_return(true)
+        allow_any_instance_of(CurationActivity).to receive(:submit_to_stripe).and_return(true)
+        allow_any_instance_of(CurationActivity).to receive(:submit_to_datacite).and_return(true)
+      end
+
+      it 'detects withdrawn state' do
+        res = @identifier.resources.last
+        res.curation_activities << CurationActivity.create(status: 'withdrawn', user: @user)
+        expect(@identifier.calculated_pub_state).to eq('withdrawn')
+      end
+
+      it 'detects last published state' do
+        resources = @identifier.resources
+        resources[1].curation_activities << CurationActivity.create(status: 'published', user: @user)
+        expect(@identifier.calculated_pub_state).to eq('published')
+      end
+
+      it 'detects last embargoed state' do
+        resources = @identifier.resources
+        resources[0].curation_activities << CurationActivity.create(status: 'embargoed', user: @user)
+        expect(@identifier.calculated_pub_state).to eq('embargoed')
+      end
+
+      it 'detects unpublished' do
+        expect(@identifier.calculated_pub_state).to eq('unpublished')
+      end
+
+      it 'handles no curation activities' do
+        @identifier.resources.each do |res|
+          res.curation_activities.destroy_all
+        end
+        expect(@identifier.calculated_pub_state).to eq('unpublished')
+      end
+
+      it 'handles no resources' do
+        @identifier.resources.destroy_all
+        expect(@identifier.calculated_pub_state).to eq('unpublished')
+      end
+
+    end
+
+    # because it overrides the ActiveRecord one
+    describe '#pub_state' do
+
+      before(:each) do
+        # a way to neuter all the callback activity
+        allow_any_instance_of(CurationActivity).to receive(:update_solr).and_return(true)
+        allow_any_instance_of(CurationActivity).to receive(:submit_to_stripe).and_return(true)
+        allow_any_instance_of(CurationActivity).to receive(:submit_to_datacite).and_return(true)
```

Comment:
Good point.  I could replace the first line with that.  I mostly don't want to trigger callbacks at all when I'm setting things up for a test case since I'm not testing the callbacks during setup, but a certain state and the specific methods without entangling myself with callbacks during setup. I don't care what the callbacks do since I'm not testing them and just don't want them to get in my way while trying to test something else.

It may be that we should have a more global way to do that.  It is also not a great help that our code is spread across different engines which are theoretically separate, but not in most practical ways, so it is harder to reuse the code.